juju/juju
# [JUJU-3559] Forward slow queries to slow log

The following forwards all the slow log queries to an independent slow log file. This should elevate the problem when you see a lot of runtime stack traces in the normal log messages. As the runtime stack traces look like panics, it has a tendency to be misread from the operator's perspective.

The slow logger is specific to the database's slow logs. The logger is a lockless file writer, that serializes all reads and periodically, flushes them to the disk. Upon exiting from the logger, an extra sync is performed before the close.

The format of the logs is not currently parsable, which might be a worthy change later on. If we can gather more tooling around slow logs, it might be something we automatically turn on in CI? For now, the log template is the following:

```
slow query took %0.3fs for statement: %s
stack trace:
%s

```

## Checklist


## QA steps

```sh
$ juju bootstrap lxd test --build-agent
$ juju controller-config query-tracing-enabled=true
$ juju controller-config query-tracing-threshold=1ms
$ juju debug-log -m controller
```

You should still expect to see slow logs in the controller log, but the data should now be minimal

```
machine-0: 17:23:26 WARNING juju.worker.dbaccessor slow query: 0.011s request exec: "COMMIT"
```

To see the full log, you need to do the following:

```sh
$ lxc exec <instance-id> -- bash
$ tail -f /var/log/juju/slow-query.log
```
